### PR TITLE
Cov x509svid

### PR DESCRIPTION
**Pull request by otaviolcs3**

### CODE REVIEW

File: svid/x509svid/tests/check_source.c [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+END_TEST
+
+START_TEST(test_x509svid_SourceFromSVID) {}
+END_TEST
+
+START_TEST(test_x509svid_SourceFromSource)
+{
+    /// WARNING: temporary solution
+    x509svid_Source *source
+        = x509svid_SourceFromSource((workloadapi_X509Source *) 0x1);
```
Comment (tjamir) [inline svid/x509svid/tests/check_source.c:14]: What does this 0x1 mean?
I suggest using a `#define` with a meaningful name.

File: svid/x509svid/tests/check_source.c [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#include "svid/x509svid/src/source.h"
+#include <check.h>
+
+START_TEST(test_x509svid_Source_GetX509SVID) {}
+END_TEST
+
+START_TEST(test_x509svid_SourceFromSVID) {}
+END_TEST
+
+START_TEST(test_x509svid_SourceFromSource)
```
Comment (tjamir) [inline svid/x509svid/tests/check_source.c:10]: Add some comment for each test (preconditions, function under test and postconditions for example)

File: svid/x509svid/tests/check_verify.c [hunk 3]
```diff
@@ -63,11 +63,14 @@ START_TEST(test_x509svid_Verify)
     arrput(certs, leaf);
     arrput(certs, inter);
 
-    /// TODO: build bundle source
```
Comment (tjamir) [inline svid/x509svid/tests/check_verify.c:66]: I know this is not in the scope of this PR, but you could add some comments on each test case (or at least, in the test cases your are working in).

File: svid/x509svid/tests/check_svid.c [hunk 4]
```diff
@@ -284,84 +302,154 @@ END_TEST
...
     x509svid_validateKeyUsage(cert, &err);
 
     ck_assert_uint_eq(err, NO_ERROR);
 
-    BIO_free(bio_mem);
     X509_free(cert);
 }
 END_TEST
 
+START_TEST(test_x509svid_SVID_GetX509SVID)
```
Comment (tjamir) [inline svid/x509svid/tests/check_svid.c:322]: I suggest you add some comments here.

File: svid/x509svid/tests/check_svid.c [hunk 5]
```diff
@@ -174,94 +169,119 @@ END_TEST
...
+
+    ck_assert_uint_ne(err, NO_ERROR);
+    ck_assert_ptr_eq(id.td.name, NULL);
+    ck_assert_ptr_eq(id.path, NULL);
+
+    X509_free(cert);
 }
 END_TEST
 
 START_TEST(test_x509svid_validateSigningCertificates)
```
Comment (tjamir) [inline svid/x509svid/tests/check_svid.c:269]: You could add some comments here

File: svid/x509svid/tests/check_svid.c [hunk 6]
```diff
@@ -174,94 +169,119 @@ END_TEST
...
     }
     spiffeid_ID_Free(&id);
+
+    id = x509svid_validateCertificates(NULL, &err);
+
+    ck_assert_uint_ne(err, NO_ERROR);
+    ck_assert_ptr_eq(id.td.name, NULL);
+    ck_assert_ptr_eq(id.path, NULL);
 }
 END_TEST
```
Comment (tjamir) [inline svid/x509svid/tests/check_svid.c:212]: You could add some comments here

File: svid/x509svid/tests/check_svid.c [hunk 7]
```diff
@@ -174,94 +169,119 @@ END_TEST
 
 START_TEST(test_x509svid_validateCertificates)
```
Comment (tjamir) [inline svid/x509svid/tests/check_svid.c:171]: You could add some comments here

File: svid/x509svid/tests/check_svid.c [hunk 8]
```diff
@@ -115,56 +118,48 @@ START_TEST(test_x509svid_ParseRaw)
     }
     arrfree(certs);
     EVP_PKEY_free(pkey);
-    BIO_free(bio_mem);
     x509svid_SVID_Free(svid);
 }
 END_TEST
 
 START_TEST(test_x509svid_newSVID)
```
Comment (tjamir) [inline svid/x509svid/tests/check_svid.c:126]: You could add some comments here

File: svid/x509svid/tests/check_svid.c [hunk 9]
```diff
@@ -10,15 +10,25 @@ START_TEST(test_x509svid_Load)
...
+
+    ck_assert_uint_ne(err, NO_ERROR);
+    ck_assert_ptr_eq(svid, NULL);
+
+    svid = x509svid_Load("./resources/good-leaf-and-intermediate.pem", NULL, &err);
+
+    ck_assert_uint_ne(err, NO_ERROR);
+    ck_assert_ptr_eq(svid, NULL);
 }
 END_TEST
```
Comment (tjamir) [inline svid/x509svid/tests/check_svid.c:35]: You could add some comments here

**Review by tjamir — CHANGES_REQUESTED** [review]

File: svid/x509svid/src/svid.c [hunk 10]
```diff
@@ -110,31 +110,25 @@ x509svid_SVID *x509svid_newSVID(X509 **certs, EVP_PKEY *pkey, err_t *err)
 
 spiffeid_ID x509svid_validateCertificates(X509 **certs, err_t *err)
 {
-    if(certs) {
-        if(arrlenu(certs) > 0) {
-            spiffeid_ID leaf = x509svid_validateLeafCertificate(certs[0], err);
-            if(!(*err)) {
-                // leaf certified
-                X509 *leaf_cert = certs[0];
+    if(arrlenu(certs) > 0) {
```
[the remark this answers is not in the document]
Comment (otaviolcs3) [reply svid/x509svid/src/svid.c:113]: Yes, arrlenu checks for NULL pointer already.

File: svid/x509svid/src/svid.c [hunk 11]
```diff
@@ -352,8 +346,8 @@ void x509svid_SVID_Free(x509svid_SVID *svid)
 
 x509svid_SVID *x509svid_SVID_GetDefaultX509SVID(x509svid_SVID **svids)
 {
-    if(!svids) {
-        return NULL;
+    if(arrlenu(svids) > 0) {
```
[the remark this answers is not in the document]
Comment (otaviolcs3) [reply svid/x509svid/src/svid.c:349]: Same answer.

File: svid/x509svid/tests/check_source.c [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+{
+    /// WARNING: temporary solution
```
[the remark this answers is not in the document]
Comment (otaviolcs3) [reply svid/x509svid/tests/check_source.c:12]: Issue created.

File: bundle/CMakeLists.txt [hunk 13]
```diff
@@ -36,17 +34,6 @@ install(
     DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/${TARGET_NAME}/jwtbundle/src"
 )
 
-# Install Headers:
```
Comment (rodrigolc) [inline bundle/CMakeLists.txt:39]: OK

File: svid/x509svid/src/svid.h [hunk 14]
```diff
@@ -102,14 +102,14 @@ bool x509svid_keyMatches(EVP_PKEY *priv_key, EVP_PKEY *pub_key, err_t *err);
 /**
  * Frees a X509-SVID object.
  *
- * \param svid [in] -SVID object pointer.
+ * \param svid [in] SVID object pointer.
```
Comment (rodrigolc) [inline svid/x509svid/src/svid.h:105]: OK

File: svid/x509svid/tests/CMakeLists.txt [hunk 15]
```diff
@@ -42,3 +42,23 @@ target_link_libraries(check_x509verify svid ${CHECK_LIBRARIES}
   pthread)
 
 add_test(check_x509verify check_x509verify)
+
+add_executable(check_x509svid_source check_source.c)
```
Comment (rodrigolc) [inline svid/x509svid/tests/CMakeLists.txt:46]: OK

File: svid/x509svid/src/svid.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (rodrigolc) [reply svid/x509svid/src/svid.c:113]: seems ok

File: svid/x509svid/tests/check_svid.c [hunk 16]
```diff
@@ -37,13 +54,13 @@ START_TEST(test_x509svid_Parse)
...
+    ck_assert_ptr_ne(svid, NULL);
     ck_assert_uint_eq(arrlenu(svid->certs), ITERS);
-    ck_assert(svid->id.td.name != NULL);
-    ck_assert(svid->id.path != NULL);
+    ck_assert_ptr_ne(svid->id.td.name, NULL);
+    ck_assert_ptr_ne(svid->id.path, NULL);
     ck_assert_str_eq(svid->id.td.name, "example.org");
     ck_assert_str_eq(svid->id.path, "/workload-1");
-    ck_assert(svid->private_key != NULL);
+    ck_assert_ptr_ne(svid->private_key, NULL);
```
Comment (rodrigolc) [inline svid/x509svid/tests/check_svid.c:64]: needs testing for pemutil_ParseCertificates returning an error inside x509svid_Parse, so we cover freeing certs

File: svid/x509svid/tests/check_svid.c [hunk 17]
```diff
@@ -56,33 +73,26 @@ START_TEST(test_x509svid_ParseRaw)
     const int ITERS = 4;
 
     FILE *f = fopen("./resources/good-leaf-and-intermediate.pem", "r");
```
Comment (rodrigolc) [inline svid/x509svid/tests/check_svid.c:75]: same as above

File: svid/x509svid/tests/check_svid.c (hunk 16 above)
Comment (rodrigolc) [reply svid/x509svid/tests/check_svid.c:64]: also, needs testing for passing null values

File: svid/x509svid/tests/check_svid.c [hunk 18]
```diff
@@ -115,56 +125,50 @@ START_TEST(test_x509svid_ParseRaw)
...
+    ck_assert_ptr_ne(svid, NULL);
     ck_assert_uint_eq(arrlenu(svid->certs), ITERS);
-    ck_assert(svid->id.td.name != NULL);
-    ck_assert(svid->id.path != NULL);
+    ck_assert_ptr_ne(svid->id.td.name, NULL);
+    ck_assert_ptr_ne(svid->id.path, NULL);
     ck_assert_str_eq(svid->id.td.name, "example.org");
     ck_assert_str_eq(svid->id.path, "/workload-1");
-    ck_assert(svid->private_key != NULL);
+    ck_assert_ptr_ne(svid->private_key, NULL);
```
Comment (rodrigolc) [inline svid/x509svid/tests/check_svid.c:170]: needs checking for passing null for certs and pkey

File: svid/x509svid/tests/check_svid.c [hunk 19]
```diff
@@ -174,94 +178,126 @@ END_TEST
...
+    id = x509svid_validateCertificates(NULL, &err);
+
+    ck_assert_uint_ne(err, NO_ERROR);
+    ck_assert_ptr_eq(id.td.name, NULL);
+    ck_assert_ptr_eq(id.path, NULL);
 }
 END_TEST
 
 START_TEST(test_x509svid_validateLeafCertificate)
 {
```
Comment (rodrigolc) [inline svid/x509svid/tests/check_svid.c:226]: Needs testing for supplying a CA certificate, which should fail

File: svid/x509svid/tests/check_svid.c [hunk 20]
```diff
@@ -174,94 +178,126 @@ END_TEST
...
-    string_t buffer = FILE_to_string(f);
-    fclose(f);
-
-    BIO *bio_mem = BIO_new(BIO_s_mem());
-    BIO_puts(bio_mem, buffer);
-    arrfree(buffer);
+    FILE *f = fopen("./resources/good-leaf-and-intermediate.pem", "r");
+    ck_assert_ptr_ne(f, NULL);
 
     X509 **certs = NULL;
```
Comment (rodrigolc) [inline svid/x509svid/tests/check_svid.c:290]: needs testing for non-CA certs, and certs with no signature flag

File: svid/x509svid/tests/check_svid.c [hunk 21]
```diff
@@ -284,84 +318,154 @@ END_TEST
 START_TEST(test_x509svid_validateKeyUsage)
 {
```
Comment (rodrigolc) [inline svid/x509svid/tests/check_svid.c:319]: same as above wrt no signature flag, as well as no KU_CRL_SIGN flag

File: svid/x509svid/tests/check_svid.c [hunk 22]
```diff
@@ -284,84 +318,154 @@ END_TEST
...
+
+    x509svid_SVID_Free(svid);
+}
+END_TEST
+
 START_TEST(test_x509svid_validatePrivateKey)
 {
     FILE *f = fopen("./resources/good-cert-and-key.pem", "r");
-    string_t buffer = FILE_to_string(f);
+
```
Comment (rodrigolc) [inline svid/x509svid/tests/check_svid.c:362]: needs testing for null private key, and diverging typed / unsupported keys

File: svid/x509svid/src/svid.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (rodrigolc) [reply svid/x509svid/src/svid.c:349]: needs testing for empty array

**Review by rodrigolc — CHANGES_REQUESTED** [review]
small changes requested, mostly passing null/empty values

**Review by rodrigolc — APPROVED** [review]

**Review by rodrigolc — APPROVED** [review]

**Review by tjamir — APPROVED** [review]
